### ISSUE TITLE
Type of arrays and other collection types

### ISSUE THREAD

**Issue by kaleidawave** [kaleidawave/ezno#34]
Given an array like
```ts
let x = [1, 4, "item"]
```
It is typed (or will be as `{ [0]: 1, [1]: 4, [2]: "item" }`). The problem is how to reference its prototype. 

Array is a generic structure so it could be `Array<1 | 4 | "item">`. However it would have to be modified every push...

It is needed to test equality
```ts
let x: Array<number> = [1, 4, "item"]
```

Alternatively the `T` type could be figured out from items. Along the lines of `interface Array<T is this[number]>`...

This also affects `Set`, `Map`...

**Comment by kaleidawave** [comment]
Won't go into too much detail here but
- Lookup generics have solved type parameters of collection types
- User created `Promise` is a `Type::SpecialObject` and will get a special implementation

Still to implement before closing
- Lookup generics for Set and Map

**Comment by PatrickLaflamme** [comment]
@kaleidawave - Is this covered in your currently open PRs? If not I'd like to take a stab at it.

**Comment by kaleidawave** [comment]
Yeah I hopefully will get `Set`, `Map` and `Promise` type structure working in #146. After that this issue should be finished

**Comment by kaleidawave** [comment]
Didn't happen in #146.

There exists a mechanism called "lookup/computed generics". It [works for subtyping](https://github.com/kaleidawave/ezno/blob/main/checker/specification/specification.md#array-property-checking) and [other places](https://github.com/kaleidawave/ezno/blob/main/checker/specification/specification.md#computed-generics-from-collection). 

Lookup generics is manually set, it would be good if it could infer for any array (via trying to find a member, preferable contravariant else covariant for some edge cases).

Will test for `Map` & `Set` soon.

> Proxy is a different issue #33 and Promises are different #168
